updates notification drawer to support mobile

matthewcarleton opened the pull request:
## Description
The purpose of this is to add support for the notification drawer on vertical nav for mobile.

We will still need to address the issue with the jQuery where to drawer does not close when the hamburger menu is selected. **This was fixed with Jeff's PR below.**

This is related to this story:

https://patternfly.atlassian.net/browse/PTNFLY-1461

* On smaller screens I've added a full height.
* I removed the expansion option on the drawer on smaller screens.
* In order to have the drawer sit above the menu on smaller screens I've introduced a unique z-index of 1035.

## Link to rawgit and/or image

Vertical Example
https://rawgit.com/matthewcarleton/patternfly/notification-drawer-mobile-dist/dist/tests/notification-drawer-vertical-nav.html

Horizontal Example
https://rawgit.com/matthewcarleton/patternfly/notification-drawer-mobile-dist/dist/tests/notification-drawer-horizontal-nav.html


## PR checklist (if relevant)

- [x] **Responsive**: works in extra small, small, medium and large view ports.
- [x] **Preview the PR**: An image or a URL for designer to preview this PR is provided.

jeff-phillips-18 commented on src/less/navbar-vertical.less:
```diff
@@ -18,7 +18,7 @@
     right: 0;
     top: 0;
-    z-index: @zindex-navbar-fixed;
+    z-index: 1035;
```
This is a little concerning.  Lots of the vertical-nav.less classes use  @zindex-navbar-fixed as a base and add to that ( https://github.com/patternfly/patternfly/blob/master/src/less/vertical-nav.less#L293-L298 ). Adopters may also be using this to be able to show items over the navbar.

Is there sufficient reason to change this?

matthewcarleton replied:
@jeff-phillips-18 we can revert it back now that we have the hamburger menu closing the drawer. This was changed to ensure the drawer sits about the open menu but that shouldn't be a problem.

matthewcarleton replied:
@jeff-phillips-18 ok, the issue is now. If I change it back to 1030 then the menu can be open and if you open the drawer it's under the menu. Does it make sense to have the menu close if you open the drawer?

jeff-phillips-18 replied:
Yeah, it should close

jeff-phillips-18 commented on src/less/notifications-drawer.less:
```diff
@@ -201,6 +201,12 @@
+      @media (max-width: @screen-xs-max) {
+        width:100%;
+        height: 100vh; // override to show use full height
+      }
```
Should the top value (58px) be subtracted?

jeff-phillips-18 replied:
it is required for the drawer to expand/collapse correctly and still show the last group or actions/notifications if there is more than fit on the page.

matthewcarleton replied:
ah ok

jeff-phillips-18 commented on tests/pages/_includes/widgets/notification-drawer.html:
```diff
@@ -1,6 +1,6 @@
   <div class="drawer-pf-title">
-    <a class="drawer-pf-toggle-expand fa fa-angle-double-left"></a>
+    <a class="drawer-pf-toggle-expand fa fa-angle-double-left hidden-xs"></a>
     <a  class="drawer-pf-close pficon pficon-close"></a>
```
We don't show the expander at xs resolution but the drawer could still have the drawer-pf-expanded class (if expanded prior to shrinking the window). Is this something we should handle?

matthewcarleton replied:
with the change I just pushed I added in a media query to remove the left:270; that is added when expanded. It now will just take up full width. I think this is a good solution, what do you think?

jeff-phillips-18 replied:
Agreed.

matthewcarleton commented on src/less/notifications-drawer.less:
```diff
@@ -201,6 +201,12 @@
+      @media (max-width: @screen-xs-max) {
+        width:100%;
+        height: 100vh; // override to show use full height
+      }
```
@jeff-phillips-18 I'll add it in, it didn't have any visible difference when I had it but it can't hurt.

matthewcarleton commented on tests/pages/_includes/widgets/notification-drawer.html:
```diff
@@ -1,6 +1,6 @@
   <div class="drawer-pf-title">
-    <a class="drawer-pf-toggle-expand fa fa-angle-double-left"></a>
+    <a class="drawer-pf-toggle-expand fa fa-angle-double-left hidden-xs"></a>
     <a  class="drawer-pf-close pficon pficon-close"></a>
```
@jeff-phillips-18 good question. How concerned are we with these types of scenarios. They seem unlikely to me.

jeff-phillips-18 replied:
I'm OK with leaving it out. The JS for this is not shipped as an available function anyhow. We'll just need to remember to handle this in the JS repos.

matthewcarleton commented:
@jeff-phillips-18 I've updated this to reflect the changes requested in https://github.com/patternfly/angular-patternfly/issues/576 by @benjaminapetersen. Also added the js provided.

benjaminapetersen commented:
Relevant to web console: https://github.com/openshift/origin-web-console/pull/1326

benjaminapetersen commented on src/less/notifications-drawer.less:
```diff
@@ -42,7 +42,9 @@
   }
   &.drawer-pf-expanded {
-    left: 270px;
+    @media (min-width: @screen-xs-max) {
```
Reference screenshot re: should it go full width on mobile:

![pf-drawer-not-full-mobile](https://user-images.githubusercontent.com/280512/29724296-fc567048-8995-11e7-944e-566298f5e00f.png)

matthewcarleton replied:
You are looking at the horizontal example which hasn't been fixed yet, this PR is for the vertical example. I am working on the horizontal separately.

matthewcarleton commented:
I've updated this to support the horizontal menu as well. I'm wondering if showing the notification icon while showing the drawer on mobile makes sense. It will offer context when screen is sized down as well as show update when all notifications are read.

benjaminapetersen commented on src/less/notifications-drawer.less:
```diff
@@ -220,6 +235,11 @@
+    @media (max-width: @screen-xs-max) {
+      width:100%;
+      height: calc(~"100vh - 58px");
+      top:34px; // this pushed the drawer below the notification icon in the menu
```
Curious if the `top: 34px` could be a variable that we could build off of.  I know we have had conversations in the web console about having things pile up in that top left corner, once you open the drawer & have a toast or two... @rhamilto thoughts?

srambach commented on src/less/notifications-drawer.less:
```diff
@@ -201,6 +210,12 @@
+
+      @media (max-width: @screen-xs-max) {
+        width:100%;
+        height: calc(~"100vh - 58px");
```
I'd love to see the px measurements in variables, especially since they are used multiple times and height relates to top.

benjaminapetersen replied:
Yup, esp if there is relationships between these values across components.

rhamilto commented:
>Curious if the top: 34px could be a variable that we could build off of. I know we have had conversations in the web console about having things pile up in that top left corner, once you open the drawer & have a toast or two... @rhamilto thoughts?

I think all of these positioning values should be variables so they can easily be overridden downstream.  Otherwise entire rules have to be duplicated...

matthewcarleton commented:
@srambach @benjaminapetersen @rhamilto That makes a lot of sense. @LHinson should we pull @andresgalante into this when he is back?

jeff-phillips-18 commented on src/js/patternfly-functions-vertical-nav.js:
```diff
@@ -259,6 +261,15 @@
+              }
             }
           } else if (navElement.hasClass('collapsed')) {
             window.localStorage.setItem('patternfly-navigation-primary', 'expanded');
```
This was done specifically for an application that needs this state maintained between restarts.

benjaminapetersen replied:
Hmm. possible to consider making these kinds of things configurable so that the needs of one app don't have to affect another?

jeff-phillips-18 replied:
Sure, this seems separate from this PR though. Please file an issue and consider contributing a PR.

benjaminapetersen replied:
Yup, not relevant for this PR.  Will do!

LHinson commented:
Thanks @srambach @rhamilto @benjaminapetersen for the reviews. After syncing with @matthewcarleton, we are in agreement with your recommendation to make the position valuables variables. 

I've filed two new issues for this. One for the notification drawer #725 and the other for the toast notifications #726.  

Considering this update out of scope for this particular PR however will address that update next.

benjaminapetersen commented on src/less/navbar-vertical.less:
```diff
@@ -46,7 +46,7 @@
           color: @navbar-pf-vertical-active-color;
         }
       }
```
Quick nit, you prob could eliminate this file from the PR as its just a whitespace change, or make sure you are using the `patternfly/~.editorconfig`.  If everyone contributing is using it little whitespace shuffles like this should disappear.

benjaminapetersen commented:
Pulled these updates into my local running `origin-web-console`.  Noticed the following issue:

![2017-08-29 10 41 53](https://user-images.githubusercontent.com/280512/29827050-e9aba15c-8ca6-11e7-9dc4-70a954c47b0d.gif)

I believe you are switching to a mobile layout too soon:

![screen shot 2017-08-29 at 10 37 43 am](https://user-images.githubusercontent.com/280512/29827025-d8ff15aa-8ca6-11e7-8683-d934f6ebae2d.png)

matthewcarleton commented:
@benjaminapetersen curious to see how this is built. Can you show me the markup for one of the notifications?

spadgett commented:
@matthewcarleton We're using angular-patternfly, which is handling the markup.

@jeff-phillips-18 Are any changes are needed downstream in angular-patternfly for this PR?

jeff-phillips-18 commented:
@spadgett I think it is the notification body include that Matt is interested in. Angular Patternfly does not set the html layout of the notification body.

spadgett commented:
Ah right, understood. Here is the markup structure:

https://github.com/openshift/origin-web-console/blob/master/app/views/directives/notifications/notification-body.html

matthewcarleton commented:
@spadgett ah ok. So from what I can see you've changed the structure quite a bit and that's going to effect how the areas react at the breakpoint. I can create a quick html prototype to show what you'd need to get it to work if that is helpful.

benjaminapetersen commented:
I am somewhat interested in the reason to move away from a `container-fluid` with rows/cols inside that use `col-md*`, `col-sm*`, etc to [adjust the layout](https://getbootstrap.com/docs/3.3/css/#grid-example-mixed-complete).  Seems like that might be a good, generic but bootstrap consistent way to solve this?

spadgett commented:
@matthewcarleton What is the normal structure? We can try to align if that's the right thing to do and it accommodates the design.

This might break other downstream projects, though, since angular-patternfly allows any markup for the notification body. It seems like the CSS is assuming a lot about what the content is. Chaning flex-direction from row to column is significant :)

We've had similar problems in the past upgrading Patternfly with components like list-view.

matthewcarleton commented:
@spadgett @benjaminapetersen have a look at the following: http://recordit.co/Lf1wsgXM4Y

I've followed the example code here http://www.patternfly.org/pattern-library/communication/notification-drawer/#/code

I've added an additional 
`<span class="drawer-pf-notification-message">Your additional content here</span>`
after the initial:
`<span class="drawer-pf-notification-message">A New Event! Huzzah! Bold!</span>`

This will allow you to make a simple change but stick to what patternfly suggests.

LHinson commented:
Thanks @matthewcarleton for providing that. 

Related to this issue and from a PF perspective, I think that an expanded view of the notification drawer (with additional content) is a new requirement for PatternFly and requires design review and docs. 

Looping in @serenamarie125 as FYI who will help determine the priority for this. I've filed an issue against the patternfly-design repo regarding this issue to kick start a conversation: https://github.com/patternfly/patternfly-design/issues/405

Since the above it out of scope, I'm going to merge this one and we will continue to follow up on the wizard as needed.

benjaminapetersen commented:
Cross referencing some further updates @sg00dwin has done in the openshift web console to help improve readability with the type of content we are handling.

PR includes some discussion to upstream a few items.

https://github.com/openshift/origin-web-console/pull/1997
